File: server/models/student_preferences.go
package models

import (
	surrealModels "github.com/surrealdb/surrealdb.go/pkg/models"
)

type StudentPreferencesEntity struct {
	ID                  surrealModels.RecordID `json:"id" swaggertype:"object"`
	PrivacyPreferences  PreferencesVisibility  `json:"privacyPreferences" validate:"required"`
	SchedulePreferences SchedulePreferences    `json:"schedulePreferences" validate:"required"`

	Student surrealModels.RecordID `json:"student" swaggertype:"object"`
}

type SchedulePreferences struct {
	DefaultOrder        string               `json:"default_order" validate:"required"`
	PreferredSchedules  []SchedulePreference `json:"preferred_schedules" validate:"required"`
	ProhibitedSchedules []SchedulePreference `json:"prohibited_schedules" validate:"required"`
}

type SchedulePreference struct {
	DayOfWeek      int `json:"day_of_week" validate:"required,min=0,max=6"`
	StartingHour   int `json:"starting_hour" validate:"min=0,max=23"`
	StartingMinute int `json:"starting_minute" validate:"min=0,max=59"`
	EndingHour     int `json:"ending_hour" validate:"min=0,max=23"`
	EndingMinute   int `json:"ending_minute" validate:"min=0,max=59"`
}

type Visibility string

const (
	VisibilityPublic         Visibility = "public"
	VisibilityFriendsFriends Visibility = "friendsFriends"
	VisibilityOnlyFriends    Visibility = "onlyFriends"
	VisibilityPrivate        Visibility = "private"
)

type PreferencesVisibility struct {
	ShowFriends  Visibility `json:"show_friends" validate:"required,oneof=public friendsFriends onlyFriends private"`
	ShowSchedule Visibility `json:"show_schedule" validate:"required,oneof=public friendsFriends onlyFriends private"`
	ShowSubjects Visibility `json:"show_subjects" validate:"required,oneof=public friendsFriends onlyFriends private"`
}

File: client/src/interfaces/preferences/StudentPreferences.ts
import type { Id } from '../surrealDb';
import type { DatabaseSchedule } from '../common/time';

export type Visibility = 'public' | 'friendsFriends' | 'onlyFriends' | 'private';

export enum OrderBySubjectOffers {
  Alphabetical = 'alphabetical',
  Difficulty = 'avg_difficulty',
  Grade = 'avg_grade',
  Workload = 'avg_workload',
  Prelations = 'prelations',
  Friends = 'differentFriends',
}



export interface SchedulePreferences {
  default_order: OrderBySubjectOffers;
  preferred_schedules: DatabaseSchedule[];
  prohibited_schedules: DatabaseSchedule[];
}

export interface PrivacyPreferences {
  show_friends: Visibility;
  show_schedule: Visibility;
  show_subjects: Visibility;
}

export interface StudentPreferencesEntity {
  id: Id<'student_preferences'>;
  student: Id<'students'>;

  privacyPreferences: PrivacyPreferences;
  schedulePreferences: SchedulePreferences;
}
